### PR TITLE
RavenDB-16999- Creation of databases and indexes implementation.  

### PR DESCRIPTION
**Pull request by omerlewitz**
**Issue link**
https://issues.hibernatingrhinos.com/issue/RavenDB-16998
https://issues.hibernatingrhinos.com/issue/RavenDB-16999

**Additional description**
Include details of the change made in this Pull Request or additional notes for the solution. Anything that can be useful for reviewers of this PR.

Please delete below the options that are not relevant

**Type of change**
-  Fixed `parseData` failure when using the secured mode.
- `OpenZipFile` remained the same. The change we discussed didn't work so I decided we will keep it this way if it's not major.
- ReadME file updated.
- Database `key` is saved after `hashing`.
- `d.HasChange` does the same thing as `d.get` in my inspection. do we want to remove it ?
- `var errAllDown *ravendb.AllTopologyNodesDownError` sometimes is null. so I had to add `passive` case for initial creation of the nodes.
- `string(index + 'A')` replaced due to warning. doe's the same thing but it's more idiomatic.
- I'm using from now on the `cast` library. it's encapsulated and much cleaner.
-  Studio froze when I was trying to reach `Database Settings` and all of the settings in the same window tab. 
   see https://issues.hibernatingrhinos.com/issue/RavenDB-17777
   This change -  `settings["Security.AuditLog.RetentionTimeInHours"] = "52560"` fixed it.

**BUG**
When creating a database with replication nodes in some of the execution I'm gettings creation in different nodes than what I was expecting.
e.g `replication_nodes = ["B", "C"]` sometimes will be created in nodes - `"C" ,"A" `.
I was checking the database record when debugging and the request was fine although the result was different.
When using secured more and distribute the database `key` it results in failure because the creation is done on different nodes.
 Any idea what can cause it ?

### CODE REVIEW

File: ravendb/resource_ravendb_server.go [hunk 1]
```diff
@@ -358,9 +655,143 @@ func parseData(d *schema.ResourceData) (ServerConfig, error) {
...
+
+		for _, db := range databasesSlice {
+			val = cast.ToStringMap(db)
+			name := cast.ToString(val["name"])
+			key := cast.ToString(val["encryption_key"])
+			if len(strings.TrimSpace(key)) != 0 && sc.Unsecured == true {
+				return nil, errors.New("encryption key can be used only in secured mode. ")
+			}
+			if sc.Unsecured == true {
+			}
```
Comment (ayende) [inline ravendb/resource_ravendb_server.go]: ??

File: ravendb/resource_ravendb_server.go [hunk 2]
```diff
@@ -358,9 +655,143 @@ func parseData(d *schema.ResourceData) (ServerConfig, error) {
...
+			databaseSettings := cast.ToStringMapString(val["settings"])
+			replicationNodes := cast.ToStringSlice(val["replication_nodes"])
+			if len(replicationNodes) == 0 {
+				replicationNodes[0] = "A"
+			}
```
Comment (ayende) [inline ravendb/resource_ravendb_server.go]: That will lead to invalid index error, no? You are verifying no elements, then set the first one?

File: ravendb/resource_ravendb_server.go [hunk 3]
```diff
@@ -476,6 +911,67 @@ func resourceServerRead(ctx context.Context, d *schema.ResourceData, meta interf
...
+					"indexes":           flattenIndexes(db.Indexes),
+				},
+			},
+		}
+		tfs = append(tfs, tf)
+	}
+	return tfs
+}
+
+func generateHash256(encryptionKey string) string {
```
Comment (ayende) [inline ravendb/resource_ravendb_server.go]: Let's use `Argon2id` instead, see: https://pkg.go.dev/golang.org/x/crypto/argon2

This is not technically needed, but better to have extra level of security here.
Also, probably need to call this `encryption_key_hash`

File: ravendb/rvn_server.go [hunk 4]
```diff
@@ -403,6 +478,9 @@ func (sc *ServerConfig) deployServer(publicIP string, index int) (err error) {
 	settings["ServerUrl.Tcp"] = "tcp://0.0.0.0:" + strconv.Itoa(sc.Url.TcpPort)
 	settings["Setup.Mode"] = "None"
 	settings["License.Path"] = "/etc/ravendb/license.json"
+	settings["Security.AuditLog.RetentionTimeInHours"] = "52560"
+
+	delete(settings, "Security.AuditLog.RetentionTimeInHrs")
```
Comment (ayende) [inline ravendb/rvn_server.go:491]: ?? Why you set and remove it?

File: ravendb/rvn_server.go [hunk 5]
```diff
@@ -901,12 +1053,130 @@ func (sc *ServerConfig) createDb(store *ravendb.DocumentStore) error {
...
+			if found == false {
+				rmDbInTags = append(rmDbInTags, tag)
+			} else {
+				members = append(members[:index], members[index+1:]...)
+			}
+		}
+		if rmDbInTags != nil {
+			operation := ravendb.NewDeleteDatabasesOperationWithParameters(&ravendb.DeleteDatabaseParameters{
+				DatabaseNames:             []string{dbRecord.DatabaseName},
+				HardDelete:                false,
```
Comment (ayende) [inline ravendb/rvn_server.go]: why false here?

**Review by ayende — CHANGES_REQUESTED** [review]

File: ravendb/resource_ravendb_server.go (hunk 2 above)
Comment (omerlewitz) [reply ravendb/resource_ravendb_server.go]: Correct.

File: ravendb/rvn_server.go (hunk 4 above)
Comment (omerlewitz) [reply ravendb/rvn_server.go:491]: `RetentionTimeInHrs` vs `RetentionTimeInHours`. 
`RetentionTimeInHrs` casues `404` in the studio as I mentioned in the PR description. 
Verifyed with @Danielle9897 and that's the reason for the change.

- Adding both of `Hrs` and `Hours` will solve it.
- Adding only `Hours` solve this as well.

Comment (Danielle9897) [reply ravendb/rvn_server.go:491]: https://issues.hibernatingrhinos.com/issue/RavenDB-17777
https://issues.hibernatingrhinos.com/issue/RavenDB-17779

File: ravendb/resource_ravendb_server.go (hunk 3 above)
Comment (omerlewitz) [reply ravendb/resource_ravendb_server.go]: done

File: ravendb/rvn_server.go [hunk 6]
```diff
@@ -1083,7 +1091,7 @@ func (sc *ServerConfig) modifyDatabase(store *ravendb.DocumentStore, dbRecord *r
 		if rmDbInTags != nil {
 			operation := ravendb.NewDeleteDatabasesOperationWithParameters(&ravendb.DeleteDatabaseParameters{
 				DatabaseNames:             []string{dbRecord.DatabaseName},
-				HardDelete:                false,
+				HardDelete:                true,
```
Comment (ayende) [inline ravendb/rvn_server.go:1094]: Shouldn't this be coming from the user?

Comment (omerlewitz) [reply ravendb/rvn_server.go:1094]: It's the situation after changes in `replication_nodes`.
e.g deploying on `[ "A" , "B", "C" ]` and we save the new configuration with only `"A"`
Is this the wrong way to handle it ?

**Review by ayende — APPROVED** [review]

**Pull request merged**